__init__.py

```diff
@@ -128,42 +131,102 @@ def _compute_caps_hash(self):
+
+            jid = msg.with_
+            state = self.get_omemo_state(account)
+            omemo_enabled = state.encryption.is_active(jid)
+
+            if omemo_enabled:
+                msg.msgtxt = '**Unencrypted** ' + msg.msgtxt
+
+                try:
+                    gui = self.ui_list[account].get(jid, None)
+                    if gui and gui.encryption_active():
+                        gui.plain_warning()
+                except:
+                    log.debug('No Ui present for ' + jid +
+                              ', Ui Warning not shown')
+
+    @log_calls('OmemoPlugin')
     def message_received(self, msg):
-        if msg.stanza.getTag('encrypted', namespace=NS_OMEMO):
+        if msg.stanza.getTag('encrypted', namespace=NS_OMEMO) and \
+                msg.mtype == 'chat':
             account = msg.conn.name
             log.debug(account + ' ⇒ OMEMO msg received')
 
             state = self.get_omemo_state(account)
             if msg.forwarded and msg.sent:
-                from_jid = str(msg.stanza.getAttr('to'))  # why gajim? why?
+                from_jid = str(msg.stanza.getTo())  # why gajim? why?
                 log.debug('message was forwarded doing magic')
             else:
-                from_jid = str(msg.stanza.getAttr('from'))
+                from_jid = str(msg.stanza.getFrom())
 
-            msg_dict = unpack_message(msg.stanza)
+            msg_dict = unpack_encrypted(msg.stanza.getTag
+                                        ('encrypted', namespace=NS_OMEMO))
             msg_dict['sender_jid'] = gajim.get_jid_without_resource(from_jid)
             plaintext = state.decrypt_msg(msg_dict)
 
             if not plaintext:
                 return
 
             msg.msgtxt = plaintext
-            msg.stanza.setBody(msg.msgtxt)
-
-            self.update_prekeys(account, msg_dict['sender_jid'])
+            # bug? there must be a body or the message gets dropped from history
+            msg.stanza.setBody(plaintext)
 
-            contact_jid = gajim.get_jid_without_resource(msg.fjid)
+            contact_jid = gajim.get_jid_without_resource(from_jid)
             if account in self.ui_list and \
                     contact_jid in self.ui_list[account]:
                 self.ui_list[account][contact_jid].activate_omemo()
             return False
-        elif msg.stanza.getTag('body'):
+
+        elif msg.stanza.getTag('body') and msg.mtype == 'chat':
             account = msg.conn.name
-            from_jid = str(msg.stanza.getAttr('from'))
+
+            from_jid = str(msg.stanza.getFrom())
             jid = gajim.get_jid_without_resource(from_jid)
-            gui = self.ui_list[account].get(jid, None)
-            if gui and gui.encryption_active():
-                gui.plain_warning()
+            state = self.get_omemo_state(account)
+            omemo_enabled = state.encryption.is_active(jid)
+
+            if omemo_enabled:
+                msg.msgtxt = '**Unencrypted** ' + msg.msgtxt
```

Comment:
After some discussion between @lovetox @iNPUTmice and me, we agreed to add it as is